<a href="https://colab.research.google.com/github/SnigdhaJanmanchi/Movie-Recommendation-Using-Collaborative-Filtering/blob/main/Movie_recommendation_using_CF.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
#Importing Necessary Libraries
import pandas as pd
from sklearn.metrics.pairwise import cosine_similarity

In [ ]:
#Reading datasets
df1 = pd.read_csv("/content/drive/MyDrive/Saved from Chrome/movies.csv")
df2 = pd.read_csv("/content/drive/MyDrive/Saved from Chrome/ratings.csv")

In [ ]:
df = df2.merge(df1, left_on='movieId', right_on='movieId', how='left')
df

userId  movieId  rating   timestamp                           title  \
0            1        1     4.0   964982703                Toy Story (1995)   
1            1        3     4.0   964981247         Grumpier Old Men (1995)   
2            1        6     4.0   964982224                     Heat (1995)   
3            1       47     5.0   964983815     Seven (a.k.a. Se7en) (1995)   
4            1       50     5.0   964982931      Usual Suspects, The (1995)   
...        ...      ...     ...         ...                             ...   
100831     610   166534     4.0  1493848402                    Split (2017)   
100832     610   168248     5.0  1493850091   John Wick: Chapter Two (2017)   
100833     610   168250     5.0  1494273047                  Get Out (2017)   
100834     610   168252     5.0  1493846352                    Logan (2017)   
100835     610   170875     3.0  1493846415  The Fate of the Furious (2017)   

                                             genres  
0       Adventure|Animation|Children|Comedy|Fantasy  
1                                    Comedy|Romance  
2                             Action|Crime|Thriller  
3                                  Mystery|Thriller  
4                            Crime|Mystery|Thriller  
...                                             ...  
100831                        Drama|Horror|Thriller  
100832                        Action|Crime|Thriller  
100833                                       Horror  
100834                                Action|Sci-Fi  
100835                  Action|Crime|Drama|Thriller  

[100836 rows x 6 columns]

In [ ]:
#Removing columns because we won't be using them
del df['timestamp']
del df['genres']

In [ ]:
df.head()

userId  movieId  rating                        title
0       1        1     4.0             Toy Story (1995)
1       1        3     4.0      Grumpier Old Men (1995)
2       1        6     4.0                  Heat (1995)
3       1       47     5.0  Seven (a.k.a. Se7en) (1995)
4       1       50     5.0   Usual Suspects, The (1995)

We are now going to transform the DF into a movie-to-user matrix, where each row represents one movie and the columns correspond to different users

In [ ]:
user_movie_matrix = pd.pivot_table(df, values = 'rating', index='movieId', columns = 'userId')
user_movie_matrix

userId   1    2    3    4    5    6    7    8    9    10   ...  601  602  603  \
movieId                                                    ...                  
1        4.0  NaN  NaN  NaN  4.0  NaN  4.5  NaN  NaN  NaN  ...  4.0  NaN  4.0   
2        NaN  NaN  NaN  NaN  NaN  4.0  NaN  4.0  NaN  NaN  ...  NaN  4.0  NaN   
3        4.0  NaN  NaN  NaN  NaN  5.0  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   
4        NaN  NaN  NaN  NaN  NaN  3.0  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   
5        NaN  NaN  NaN  NaN  NaN  5.0  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   
...      ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
193581   NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   
193583   NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   
193585   NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   
193587   NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   
193609   NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   

userId   604  605  606  607  608  609  610  
movieId                                     
1        3.0  4.0  2.5  4.0  2.5  3.0  5.0  
2        5.0  3.5  NaN  NaN  2.0  NaN  NaN  
3        NaN  NaN  NaN  NaN  2.0  NaN  NaN  
4        NaN  NaN  NaN  NaN  NaN  NaN  NaN  
5        3.0  NaN  NaN  NaN  NaN  NaN  NaN  
...      ...  ...  ...  ...  ...  ...  ...  
193581   NaN  NaN  NaN  NaN  NaN  NaN  NaN  
193583   NaN  NaN  NaN  NaN  NaN  NaN  NaN  
193585   NaN  NaN  NaN  NaN  NaN  NaN  NaN  
193587   NaN  NaN  NaN  NaN  NaN  NaN  NaN  
193609   NaN  NaN  NaN  NaN  NaN  NaN  NaN  

[9724 rows x 610 columns]

In [ ]:
user_movie_matrix = user_movie_matrix.fillna(0)
user_movie_matrix.head()

userId   1    2    3    4    5    6    7    8    9    10   ...  601  602  603  \
movieId                                                    ...                  
1        4.0  0.0  0.0  0.0  4.0  0.0  4.5  0.0  0.0  0.0  ...  4.0  0.0  4.0   
2        0.0  0.0  0.0  0.0  0.0  4.0  0.0  4.0  0.0  0.0  ...  0.0  4.0  0.0   
3        4.0  0.0  0.0  0.0  0.0  5.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
4        0.0  0.0  0.0  0.0  0.0  3.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
5        0.0  0.0  0.0  0.0  0.0  5.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   

userId   604  605  606  607  608  609  610  
movieId                                     
1        3.0  4.0  2.5  4.0  2.5  3.0  5.0  
2        5.0  3.5  0.0  0.0  2.0  0.0  0.0  
3        0.0  0.0  0.0  0.0  2.0  0.0  0.0  
4        0.0  0.0  0.0  0.0  0.0  0.0  0.0  
5        3.0  0.0  0.0  0.0  0.0  0.0  0.0  

[5 rows x 610 columns]

Calculating similarites

Similarity based methods determine the most similar objects with the highest values as it implies they live in closer neighborhoods. There are multiple similarity based metrics

Pearson’s correlation

Spearman’s correlation

Kendall’s Tau

Cosine similarity

Jaccard similarity

However, for our case we'll be using Pearson’s correlation because we have a sparse matrix, which means that most movies are not rated (have a rating of 0). hence we will center all the ratings to 0 so that the default rating becomes 0.


In [ ]:
#user-based collaborative filtering

user_user_matrix = user_movie_matrix.corr(method='pearson')
user_user_matrix

userId       1         2         3         4         5         6         7    \
userId                                                                         
1       1.000000  0.019400  0.053056  0.176920  0.120866  0.104418  0.143793   
2       0.019400  1.000000 -0.002594 -0.003804  0.013183  0.016257  0.021567   
3       0.053056 -0.002594  1.000000 -0.004556  0.001887 -0.004577 -0.005634   
4       0.176920 -0.003804 -0.004556  1.000000  0.121018  0.065719  0.100602   
5       0.120866  0.013183  0.001887  0.121018  1.000000  0.294138  0.101725   
...          ...       ...       ...       ...       ...       ...       ...   
606     0.121981  0.011299 -0.003246  0.164831  0.090158  0.047506  0.172499   
607     0.254200  0.005813  0.012885  0.115118  0.145764  0.142169  0.173293   
608     0.262241  0.032730  0.008096  0.116861  0.122607  0.137954  0.305439   
609     0.085434  0.024373 -0.002963  0.023930  0.258289  0.207124  0.084494   
610     0.098719  0.089329  0.015962  0.062523  0.040372 -0.010400  0.163775   

userId       8         9         10   ...       601       602       603  \
userId                                ...                                 
1       0.128547  0.055268 -0.000298  ...  0.066256  0.149942  0.186978   
2       0.023750 -0.003448  0.061880  ...  0.198549  0.010888 -0.004030   
3       0.001703 -0.003111 -0.005501  ...  0.000150 -0.000585  0.011211   
4       0.054235  0.002417  0.015615  ...  0.072848  0.114287  0.281866   
5       0.426576 -0.004185  0.023471  ...  0.061912  0.414931  0.095394   
...          ...       ...       ...  ...       ...       ...       ...   
606     0.081913  0.057989  0.054877  ...  0.153892  0.084208  0.224637   
607     0.178133  0.003257 -0.004809  ...  0.080034  0.187588  0.173025   
608     0.175912  0.086229  0.048373  ...  0.136316  0.174069  0.164479   
609     0.421627 -0.003937  0.014983  ...  0.029664  0.331053  0.046000   
610     0.058168  0.055265  0.087036  ...  0.200487  0.050806  0.064594   

userId       604       605       606       607       608       609       610  
userId                                                                        
1       0.056530  0.134412  0.121981  0.254200  0.262241  0.085434  0.098719  
2      -0.005345 -0.007919  0.011299  0.005813  0.032730  0.024373  0.089329  
3      -0.004822  0.003678 -0.003246  0.012885  0.008096 -0.002963  0.015962  
4       0.039699  0.065493  0.164831  0.115118  0.116861  0.023930  0.062523  
5       0.254117  0.141077  0.090158  0.145764  0.122607  0.258289  0.040372  
...          ...       ...       ...       ...       ...       ...       ...  
606     0.035251  0.106752  1.000000  0.115999  0.188354  0.052385  0.093851  
607     0.126267  0.101138  0.115999  1.000000  0.258245  0.142533  0.098518  
608     0.133734  0.144896  0.188354  0.258245  1.000000  0.109563  0.248944  
609     0.232115  0.089810  0.052385  0.142533  0.109563  1.000000  0.033713  
610     0.018502  0.072727  0.093851  0.098518  0.248944  0.033713  1.000000  

[610 rows x 610 columns]

As we can see from the above matrix, higher +ve values means more similar and -ve values mean not similar. Eg: Similarty of user 1 and user 2 is 0.019400 but similarity of user1 and user3 is 0.053056, which means User1 is more similar to User3.

Recommending Movies for UserId 2

Movie list for userid 2

In [ ]:
#Extracting top 10 similar users for User2 by sorting them in descending order based on their similarties

user_user_matrix.loc[2].sort_values(ascending=False).head(10)

userId
2      1.000000
366    0.297982
417    0.277366
378    0.273342
550    0.252051
189    0.240668
528    0.238262
461    0.237457
495    0.235147
435    0.231771
Name: 2, dtype: float64

In [ ]:
#Converting the above data into a DF and removing the user2 itself

df_2 = pd.DataFrame(user_user_matrix.loc[2].sort_values(ascending=False).head(10))
df_2 = df_2.reset_index()
df_2.columns = ['userId', 'similarity']

In [ ]:
df_2 = df_2.drop((df_2[df_2['userId'] ==2]).index)
df_2

userId  similarity
1     366    0.297982
2     417    0.277366
3     378    0.273342
4     550    0.252051
5     189    0.240668
6     528    0.238262
7     461    0.237457
8     495    0.235147
9     435    0.231771

In [ ]:
#New DF which has the similar users and rated movies

final_df = df_2.merge(df, left_on='userId', right_on='userId', how='left')
final_df

userId  similarity  movieId  rating                              title
0       366    0.297982      110     4.0                  Braveheart (1995)
1       366    0.297982      589     4.0  Terminator 2: Judgment Day (1991)
2       366    0.297982     1036     4.0                    Die Hard (1988)
3       366    0.297982     1089     5.0              Reservoir Dogs (1992)
4       366    0.297982     2028     4.0         Saving Private Ryan (1998)
..      ...         ...      ...     ...                                ...
596     435    0.231771    68954     4.0                          Up (2009)
597     435    0.231771    74458     4.5              Shutter Island (2010)
598     435    0.231771    79132     4.5                   Inception (2010)
599     435    0.231771    80463     4.5         Social Network, The (2010)
600     435    0.231771    82459     4.0                   True Grit (2010)

[601 rows x 5 columns]

We obtain a new metric (score) by combining how similar two people are and their movie rating. The basic thought is that a user who is most similar to you may suggest movies you would like best. If the most similar user gives a movie a high ranking, our algorithm will place it higher in the recommendation.

In [ ]:
final_df['score'] = final_df['similarity']*final_df['rating']
final_df

userId  similarity  movieId  rating                              title  \
0       366    0.297982      110     4.0                  Braveheart (1995)   
1       366    0.297982      589     4.0  Terminator 2: Judgment Day (1991)   
2       366    0.297982     1036     4.0                    Die Hard (1988)   
3       366    0.297982     1089     5.0              Reservoir Dogs (1992)   
4       366    0.297982     2028     4.0         Saving Private Ryan (1998)   
..      ...         ...      ...     ...                                ...   
596     435    0.231771    68954     4.0                          Up (2009)   
597     435    0.231771    74458     4.5              Shutter Island (2010)   
598     435    0.231771    79132     4.5                   Inception (2010)   
599     435    0.231771    80463     4.5         Social Network, The (2010)   
600     435    0.231771    82459     4.0                   True Grit (2010)   

        score  
0    1.191930  
1    1.191930  
2    1.191930  
3    1.489912  
4    1.191930  
..        ...  
596  0.927083  
597  1.042968  
598  1.042968  
599  1.042968  
600  0.927083  

[601 rows x 6 columns]

In [ ]:
#Creating a df for all the movies which are already watched by our target User 2

watched_df = df[df['userId'] == 2]
watched_df

userId  movieId  rating  \
232       2      318     3.0   
233       2      333     4.0   
234       2     1704     4.5   
235       2     3578     4.0   
236       2     6874     4.0   
237       2     8798     3.5   
238       2    46970     4.0   
239       2    48516     4.0   
240       2    58559     4.5   
241       2    60756     5.0   
242       2    68157     4.5   
243       2    71535     3.0   
244       2    74458     4.0   
245       2    77455     3.0   
246       2    79132     4.0   
247       2    80489     4.5   
248       2    80906     5.0   
249       2    86345     4.0   
250       2    89774     5.0   
251       2    91529     3.5   
252       2    91658     2.5   
253       2    99114     3.5   
254       2   106782     5.0   
255       2   109487     3.0   
256       2   112552     4.0   
257       2   114060     2.0   
258       2   115713     3.5   
259       2   122882     5.0   
260       2   131724     5.0   

                                                 title  
232                   Shawshank Redemption, The (1994)  
233                                   Tommy Boy (1995)  
234                           Good Will Hunting (1997)  
235                                   Gladiator (2000)  
236                           Kill Bill: Vol. 1 (2003)  
237                                  Collateral (2004)  
238  Talladega Nights: The Ballad of Ricky Bobby (2...  
239                               Departed, The (2006)  
240                            Dark Knight, The (2008)  
241                               Step Brothers (2008)  
242                        Inglourious Basterds (2009)  
243                                  Zombieland (2009)  
244                              Shutter Island (2010)  
245                  Exit Through the Gift Shop (2010)  
246                                   Inception (2010)  
247                                   Town, The (2010)  
248                                  Inside Job (2010)  
249                       Louis C.K.: Hilarious (2010)  
250                                     Warrior (2011)  
251                      Dark Knight Rises, The (2012)  
252            Girl with the Dragon Tattoo, The (2011)  
253                            Django Unchained (2012)  
254                    Wolf of Wall Street, The (2013)  
255                                Interstellar (2014)  
256                                    Whiplash (2014)  
257                                    The Drop (2014)  
258                                  Ex Machina (2015)  
259                          Mad Max: Fury Road (2015)  
260  The Jinx: The Life and Deaths of Robert Durst ...

In this Step we are removing the already watched movies because we don't want to recommend them again

In [ ]:
cond = final_df['movieId'].isin(watched_df['movieId'])
final_df.drop(final_df[cond].index, inplace = True)

In [ ]:
recommended_df = final_df.sort_values(by = 'score', ascending = False)['title'].head(10)
recommended_df = recommended_df.reset_index()
del recommended_df['index']

Final List of 10 Recommended Movies for User 2

In [ ]:
recommended_df

title
0                              Reservoir Dogs (1992)
1                                Pulp Fiction (1994)
2                               Trainspotting (1996)
3                        Seven (a.k.a. Se7en) (1995)
4                          American History X (1998)
5                        The Butterfly Effect (2004)
6  Lord of the Rings: The Return of the King, The...
7                City of God (Cidade de Deus) (2002)
8      Lord of the Rings: The Two Towers, The (2002)
9                           Beautiful Mind, A (2001)